### DISCUSSION TITLE
Redirect to Previous Page when Logged in.

### DISCUSSION THREAD

**Discussion by EhanDuan** [vercel/next.js#38608] [Help] [answered]
### Overview

I am using `NextJs`, `Chakra UI` and `Firebase` to build an webapp.

### Scene

I need to realize a feature: When people visit shared URLs, which required login, they log in or signup and **the page will redirect to the URLs.**

### Similar Question
https://github.com/vercel/next.js/discussions/13249

###  Intuition
The urls shared could be abstracted as **previous urls** since they will be requested firstly. Then, my function (already done) checked not logged in, it will redirect to **login page**.

In this case, I could pass the previous url as `query` named `from` in `URLObject`. When a user in login, I can get the `from` from the query. Then, after login, page will redirect to `from` url.  

Here is the big picture.

<img width="1209" alt="image" src="https://user-images.githubusercontent.com/75397936/178768365-54ec9d57-2d82-4c3a-b36b-d72c648bd676.png">

### Implementation

I used two wrapper functions to wrap page components to detect the user status.
```ts
// outer: just a function
// inner: functional component
export function withProtected(Component) {
  return function WithProtected(props) {
    const auth = useAuth();
    const router = useRouter();

    if (!auth.user) {
      router.replace({
        pathname: "/login",
        query: { from: router.asPath }
      });
      // console.log(router.asPath);
      return <h1>Redirect Into Login Page</h1>;
    }

    return <Component auth={auth} {...props} />;
  };
}

export function withPublic(Component, from) {
  return function WithPublic(props) {
    const auth = useAuth();
    const router = useRouter();

    console.log(from);
    console.log(auth.user);

    if (auth.user) {
      // replace to from url

      // router.replace("/practice"); last version hard coded to a certain url
      console.log(from);
      return <h1>Redirect Into Practice Page</h1>;
    }

    return <Component auth={auth} {...props} />;
  };
}

```

I used `withPublic` to wrap `login` page component. Once logged in, redirect to other page. I cleaned  the code for showing the error.

login.tsx
```ts

interface Props {}
let from: string = "/";
function SignIn({}: Props): ReactElement {
  const { user } = useAuth();
  const router = useRouter();
  console.log(router.query);
  // get which url this user is from. when logged in successfully, redirect to previous page
  from =
    router.query.from === undefined || Array.isArray(router.query.from)
      ? "/"
      : router.query.from;
  console.log(from);

  return (
    <div>
        <LoginForm />
    </div>
  );
}

console.log(from);

export default withPublic(SignIn, from);
```

#### Question 1: `withPublic` `from` Value not Correct
I printed `from` at each point of reference.  In `withProtected`, `from` value was successfully assigned using `router.asPath`.
And the value did into the `login` component.

**However, when I passed `from` to `export default withPublic(SignIn, from);`, it always kept as the `/` which I declared outside of the functional component.**

<img width="1069" alt="image" src="https://user-images.githubusercontent.com/75397936/178772020-6e59d5be-7335-407e-b16c-12875760f40f.png">

#### Question 2: How to handle If `login` -> `signup`
In a scenario, users need to create accounts first. For now, in `withProteced`, the page will directly to `login` as default. Then, user create their accounts.
<img width="608" alt="image" src="https://user-images.githubusercontent.com/75397936/178772382-8725b2e9-9084-4e95-a9c5-914b9ac4a30e.png">

My question is: if a user follow this flow:
<img width="1291" alt="image" src="https://user-images.githubusercontent.com/75397936/178773266-08748fa0-3d22-4413-9205-2a7ae951eba6.png">
Will my solution still work? Specifically, could my `from` pass to `signup` page and let `router` process to the shared link ?

If my solution is not a standard practice in the industry, feel free to point it out.

Thanks for your ideas!

**Comment by EhanDuan** [comment]
Hi, I just found I made it complicated. I got answer for **Question1**

### Solution

Instead of passing previous urls, I  can use `router.back()` to solve the issue!

However, there is a tiny thing:

1. If a user clicks the url, the code works well. ( Most users will follow this pattern )
2. But if a user copy and paste the url into url bar, and then visit. It will not go back to the `shared url`. Instead, it will go back to previous page like `Google Home` or others.

**Reply by JohnMwanyika** [reply]
Thats what I found too,, any work around on this?

**Comment by icyJoseph** [comment, marked as answer]
Hi,

I think we could simplify this a bit further.

```js
// hocs/index.js
import { useEffect } from "react";
import { useRouter } from "next/router";
import { useAuth } from "../hooks/useAuth";

export function withProtected(Component) {
  return function WithProtected(props) {
    const auth = useAuth();
    const router = useRouter();

    useEffect(() => {
      if (auth.status === "pending") return;

      if (!auth.user) {
        router.replace(
          {
            pathname: "/login",
            query: { from: encodeURIComponent(router.asPath) },
          },
          "/login"
        );
      }
    }, [auth.user, auth.status, router]);

    if (auth.status === "pending") return null;

    if (!auth.user) return <h1>Redirect Into Login Page</h1>;

    return <Component auth={auth} {...props} />;
  };
}
```

In this case I made my own `useAuth`:

```js
// hooks/useAuth.js
import { useState, useEffect, useMemo } from "react";

// A true useAuth impl takes care of caching, and revalidating etc, this is just for this example
export const useAuth = () => {
  const [user, setUser] = useState(null);
  const [status, setStatus] = useState("pending");

  useEffect(() => {
    const controller = new AbortController();
    const signal = controller.signal;
    fetch("/api/status", { signal })
      .then((res) => res.json())
      .then((data) => {
        setUser(data.user);
        setStatus("done");
      })
      .catch(() => {
        if (signal.aborted) {
          console.log("aborted status check");
        }
        setStatus("error");
      });

    return () => controller.abort;
  }, []);

  return useMemo(() => ({ user, status }), [user, status]);
};

```

And then I define, three pages:

```js
import { withProtected } from "../hocs";

function Home({ auth }) {
  return (
    <div>
      Welcome to this page <code>{auth.user}</code>
    </div>
  );
}

export default withProtected(Home);
```

```js
import { useRouter } from "next/router";
import Link from "next/link";

export default function Login() {
  const router = useRouter();
  const { query } = router;

  return (
    <>
      <form
        onSubmit={(e) => {
          e.preventDefault();
          const rawUsername = e.target.username.value || "";
          const rawPassword = e.target.username.value || "";

          const username = rawUsername.trim();
          const password = rawPassword.trim();

          if (!username || !password) return;

          fetch("/api/login", {
            method: "POST",
            body: JSON.stringify({ username, password }),
            headers: {
              "content-type": "application/json",
            },
          }).then(() => {
            router.replace(query.returnUrl || "/");
          });
        }}
      >
        <fieldset>
          <legend>Login</legend>
          <input
            type="text"
            id="username"
            name="username"
            placeholder="username"
          />
          <input
            type="password"
            id="password"
            name="password"
            placeholder="password"
          />

          <button type="submit">submit</button>
        </fieldset>
      </form>

      <p>Need an account?</p>

      <Link href={{ pathname: "/register", query }} as="/register" replace>
        <a>Register</a>
      </Link>
    </>
  );
}
```

```js
import { useRouter } from "next/router";

export default function Register() {
  const router = useRouter();
  const { query } = router;

  return (
    <form
      onSubmit={(e) => {
        e.preventDefault();
        const rawUsername = e.target.username.value || "";
        const rawPassword = e.target.username.value || "";

        const username = rawUsername.trim();
        const password = rawPassword.trim();

        if (!username || !password) return;

        fetch("/api/register", {
          method: "POST",
          body: JSON.stringify({ username, password }),
          headers: {
            "content-type": "application/json",
          },
        }).then(() => {
          router.replace(query.returnUrl || "/");
        });
      }}
    >
      <fieldset>
        <legend>Register</legend>
        <input
          type="text"
          id="username"
          name="username"
          placeholder="username"
        />
        <input
          type="password"
          id="password"
          name="password"
          placeholder="password"
        />

        <button type="submit">submit</button>
      </fieldset>
    </form>
  );
}
```

Finally, three API routes:

```js
// pages/api/login
// pages/api/register
export default function handler(req, res) {
  const { username } = req.body;

  res.setHeader(
    "set-cookie",
    `user=${username}; Max-Age=${1000 * 60 * 12};Path=/; HttpOnly`
  );

  return res.status(200).json({ name: "Hello, world!" });
}
```

A copy of the above in `api/register`,  and a status page:

```js
// pages/api/status
export default function handler(req, res) {
  if (req.cookies["user"]) {
    return res.status(200).json({ user: req.cookies["user"] });
  }

 return res.status(200).json({ user: null });
}
```

So the file structure looks like:

```shell
.
├── hocs
│   └── index.js
├── hooks
│   └── useAuth.js
├── next.config.js
├── package.json
├── pages
│   ├── _app.js
│   ├── api
│   │   ├── hello.js
│   │   ├── login.js
│   │   ├── register.js
│   │   └── status.js
│   ├── index.js
│   ├── login.js
│   └── register.js
├── public
│   ├── favicon.ico
│   └── vercel.svg
└── yarn.lock
```

This is but one flavour to do this. I'd prefer using `<form action="/api/login" method="POST">` or `GET`, and from the `api/route` redirect the user elsewhere, as shown here, https://nextjs.org/docs/api-routes/response-helpers#redirects-to-a-specified-path-or-url.

Though it is a bit more annoying to include additional data, you'd need to include a hidden input field set to the return url, but it'd work!

Anyhow...

The way I have it now, the HoC takes care of redirecting, inside a `useEffect`, I think this is the root of all evil here. To do such things as redirect, as part of the render is impure, and must be avoided. That's what `useEffect` is for.

Moreover, `replace` and `push`, from the router are asynchronous!

Notice I also use the second argument to `replace` which keeps the user from seeing the url return data, not necessary, but kinda nice.

In this case, when both `login` or `register` succeed, we send the user back to the route where they came from, even if they when `page -> login -> register`, they return to `page`.

What do you think? Can you apply this, more or less, to your case?

**Reply by EhanDuan** [reply]
Hi, thanks for your detailed answer. The idea is crystal clear!

But I might some trouble. I think it belongs to `React Hydration Error`. So I create a new issue #39005 .

I also put it here if convenient for you. Here is the post.

---
# Description

Previously, I was using `context` to wrap `_app.tsx` in order to do authentication.

`_app.tsx`
```tsx
function MyApp({ Component, pageProps }: AppProps) {
  return (
    <AuthProvider>
      <AuthStateChanged>
        <ChakraProvider>
          <CSSReset />
          <Component {...pageProps} />
        </ChakraProvider>
      </AuthStateChanged>
    </AuthProvider>
  );
}
```

The index page works well like below.
<img width="1738" alt="Corrected Layout" src="https://user-images.githubusercontent.com/75397936/180853231-61087c2c-bcdc-418f-96db-46515fcfda01.png">
---

Since I need to improve my app authentication mechanism. I decided to remove the `AuthProvider` and `AuthStateChanged`.

When I removed the outer `AuthProvider`, it still worked well. 

**But If I removed the `AuthStateChanged`. The page became like this.**
<img width="1719" alt="wrong layout" src="https://user-images.githubusercontent.com/75397936/180853655-6a1e7e14-a11b-46a3-9a64-9e75d6df4990.png">

Here is the Console result. I can make sure it came from my deletion.
<img width="1760" alt="error console" src="https://user-images.githubusercontent.com/75397936/180853879-a475ca39-8c77-412c-b96e-af90f2ba2ee8.png">

It looks like it is related to [React Hydration Error](https://nextjs.org/docs/messages/react-hydration-error).

I looked up the doc, it refers me to the specific example. Here is the nextjs example repo: [with-chakra-ui](https://github.com/vercel/next.js/tree/canary/examples/with-chakra-ui). I did not find a solution to solve my problem.


Do you have any idea about it ?

Thanks for your help!

**Comment by ianizaguirre** [comment]
@EhanDuan What tool did you use to make those pictures in your original question ? it looks really nice

These pics:
![image](https://user-images.githubusercontent.com/29879270/181347777-5f4808c6-bf58-420f-b0dd-0dba2dde6972.png)

**Reply by EhanDuan** [reply]
Hi @ianizaguirre 

I use draw.io. It is free for a single user.

**Reply by ianizaguirre** [reply]
cool, thanks for that info

**Comment by rosshabe** [comment]
> ### Overview
> I am using `NextJs`, `Chakra UI` and `Firebase` to build an webapp.
> 
> ### Scene
> I need to realize a feature: When people visit shared URLs, which required login, they log in or signup and **the page will redirect to the URLs.**
> 
> ### Similar Question
> #13249
> 
> ### Intuition
> The urls shared could be abstracted as **previous urls** since they will be requested firstly. Then, my function (already done) checked not logged in, it will redirect to **login page**.
> 
> In this case, I could pass the previous url as `query` named `from` in `URLObject`. When a user in login, I can get the `from` from the query. Then, after login, page will redirect to `from` url.
> 
> Here is the big picture.
> 
> <img alt="image" width="1209" src="https://user-images.githubusercontent.com/75397936/178768365-54ec9d57-2d82-4c3a-b36b-d72c648bd676.png">
> 
> ### Implementation
> I used two wrapper functions to wrap page components to detect the user status.
> 
> ```ts
> // outer: just a function
> // inner: functional component
> export function withProtected(Component) {
>   return function WithProtected(props) {
>     const auth = useAuth();
>     const router = useRouter();
> 
>     if (!auth.user) {
>       router.replace({
>         pathname: "/login",
>         query: { from: router.asPath }
>       });
>       // console.log(router.asPath);
>       return <h1>Redirect Into Login Page</h1>;
>     }
> 
>     return <Component auth={auth} {...props} />;
>   };
> }
> 
> export function withPublic(Component, from) {
>   return function WithPublic(props) {
>     const auth = useAuth();
>     const router = useRouter();
> 
>     console.log(from);
>     console.log(auth.user);
> 
>     if (auth.user) {
>       // replace to from url
> 
>       // router.replace("/practice"); last version hard coded to a certain url
>       console.log(from);
>       return <h1>Redirect Into Practice Page</h1>;
>     }
> 
>     return <Component auth={auth} {...props} />;
>   };
> }
> ```
> 
> I used `withPublic` to wrap `login` page component. Once logged in, redirect to other page. I cleaned the code for showing the error.
> 
> login.tsx
> 
> ```ts
> interface Props {}
> let from: string = "/";
> function SignIn({}: Props): ReactElement {
>   const { user } = useAuth();
>   const router = useRouter();
>   console.log(router.query);
>   // get which url this user is from. when logged in successfully, redirect to previous page
>   from =
>     router.query.from === undefined || Array.isArray(router.query.from)
>       ? "/"
>       : router.query.from;
>   console.log(from);
> 
>   return (
>     <div>
>         <LoginForm />
>     </div>
>   );
> }
> 
> console.log(from);
> 
> export default withPublic(SignIn, from);
> ```
> 
> ### Questions
> #### Question 1: `withPublic` `from` Value not Correct
> I printed `from` at each point of reference. In `withProtected`, `from` value was successfully assigned using `router.asPath`. And the value did into the `login` component.
> 
> **However, when I passed `from` to `export default withPublic(SignIn, from);`, it always kept as the `/` which I declared outside of the functional component.**
> 
> <img alt="image" width="1069" src="https://user-images.githubusercontent.com/75397936/178772020-6e59d5be-7335-407e-b16c-12875760f40f.png">
> 
> #### Question 2: How to handle If `login` -> `signup`
> In a scenario, users need to create accounts first. For now, in `withProteced`, the page will directly to `login` as default. Then, user create their accounts. <img alt="image" width="608" src="https://user-images.githubusercontent.com/75397936/178772382-8725b2e9-9084-4e95-a9c5-914b9ac4a30e.png">
> 
> My question is: if a user follow this flow: <img alt="image" width="1291" src="https://user-images.githubusercontent.com/75397936/178773266-08748fa0-3d22-4413-9205-2a7ae951eba6.png"> Will my solution still work? Specifically, could my `from` pass to `signup` page and let `router` process to the shared link ?
> 
> If my solution is not a standard practice in the industry, feel free to point it out.
> 
> Thanks for your ideas!

Thanks for this saved me big time...
